Userland/Libraries/LibGL/SoftwareGLContext.cpp

```diff
@@ -648,6 +777,107 @@ void SoftwareGLContext::gl_cull_face(GLenum cull_mode)
     m_culled_sides = cull_mode;
 }
 
+GLuint SoftwareGLContext::gl_gen_lists(GLsizei range)
+{
+    if (range <= 0) {
+        m_error = GL_INVALID_VALUE;
+        return 0;
+    }
+    if (m_in_draw_state) {
+        m_error = GL_INVALID_OPERATION;
+        return 0;
+    }
+
+    auto initial_entry = m_listings.size();
+    m_listings.resize(range + initial_entry);
+    return initial_entry + 1;
+}
+
+void SoftwareGLContext::gl_call_list(GLuint list)
+{
+    if (should_append_to_listing()) {
+        if (m_gl_call_depth > max_allowed_gl_call_depth)
+            return;
+
+        append_to_listing<&SoftwareGLContext::gl_call_list>(list);
+        if (!should_execute_after_appending_to_listing())
+            return;
+    }
+
+    if (m_listings.size() < list)
+        return;
+
+    TemporaryChange { m_gl_call_depth, m_gl_call_depth + 1 };
+
+    auto& listing = m_listings[list - 1];
+    for (auto& entry : listing.entries) {
+        entry.function.visit([&](auto& function) {
+            entry.arguments.visit([&](auto& arguments) {
+                auto apply = [&]<typename... Args>(Args && ... args)
+                {
+                    if constexpr (requires { (this->*function)(forward<Args>(args)...); })
+                        (this->*function)(forward<Args>(args)...);
+                };
+
+                arguments.apply_as_args(apply);
+            });
+        });
+    }
+}
```

Comment:
This is nifty!

Reply:
Just about the only nifty part of this :P